Implement configurable engine options

nathanleclaire opened the pull request:
cc @ehazlett @sthulb @hairyhenderson @ibuildthecloud PTAL

Implementation of feature spec'd out in https://github.com/docker/machine/issues/974

I have been testing with labels, storage drivers, and insecure registry and it seems to be working pretty well.  The generated configuration files aren't great looking but that's the name of the game.

Signed-off-by: Nathan LeClaire <nathan.leclaire@gmail.com>

nathanleclaire commented:
Still to go: Docs and integration tests, maybe a `--engine-arbitrary-flag` flag as an "escape hatch" when the user wants flags which are not supported directly, e.g.:

```console
$ docker-machine create -d virtualbox --engine-arbitrary-flag icc=false foobar
```

ibuildthecloud commented:
What is the expected behavior when a provisioner doesn't support an option.  Should provisioners return an error, log a warn, or ignore?

hairyhenderson commented:
Nice. This seems pretty good to me.

`--engine-arbitrary-flag` seems a little non-intuititive... Would it be possible instead to just (eventually) support all of the engine options?

Not sure if `codegangsta/cli` supports it, but if we could just take any unknown option that starts with `--engine-*` and pass it through to the engine, that might fill that need.

nathanleclaire commented:
> Should provisioners return an error, log a warn, or ignore?

With the current implementation, an error.  Then the user will have to re-create without specifying the option.

sthulb commented:
@hairyhenderson codegangsta/cli doesn't support that unfortunately, but I like @nathanleclaire's solution.

hairyhenderson commented:
@sthulb - Ah. In that case @nathanleclaire's solution is OK.

However, what about something a little shorter - `--engine-flag` or `--engine-opt` or something?

ehazlett commented:
Nice -- I'm going to test this out!  Thanks!

+1 to @hairyhenderson -- i think something perhaps `--engine-flag` or `--engine-opt` is a little more friendly than `arbitrary-flag`.

ehazlett commented on libmachine/provision/ubuntu.go:
```diff
@@ -151,18 +164,58 @@ func (provisioner *UbuntuProvisioner) CompatibleWithHost() bool {
+--tlscacert {{.AuthOptions.CaCertRemotePath}}
+--tlscert {{.AuthOptions.ServerCertRemotePath}}
+--tlskey {{.AuthOptions.ServerKeyRemotePath}}
+{{ range .EngineOptions.Labels }}
```
If you wanted to condense these to reduce whitespace in the config you can use it like this (although it would also reduce readability):

```
{{ range .EngineOptions.Labels }}--label {{.}}{{ end}}
```

nathanleclaire replied:
Yeah I keep going back and forth on that one.  I think I will go with the format that you mentioned here since the `range` statements aren't too crazy.

ehazlett commented on libmachine/provision/utils.go:
```diff
@@ -37,12 +37,36 @@ func installDockerGeneric(p Provisioner) error {
-func ConfigureAuth(p Provisioner, authOptions auth.AuthOptions) error {
+func makeDockerOptionsDir(p Provisioner) error {
+	dockerDir := p.GetDockerOptionsDir()
+	_, err := p.SSHCommand(fmt.Sprintf("sudo mkdir -p %s", dockerDir))
```
Style nitpick:

```
if _, err := p.SSHCommand(fmt.Sprintf("sudo mkdir -p %s", dockerDir)); err != nil {
```

nathanleclaire replied:
Yup I noticed this after the fact as well -- Will update

ehazlett commented:
Just a few minor comments / nitpicks -- overall looking awesome!  Can't wait to get this :+1:

nathanleclaire commented:
Updated some stylistic things, added `--engine-flag`, and updated some documentation.  Getting pretty close, just want to write a few integration tests now.

hairyhenderson commented:
:+1:

jamshid commented:
Does this feature only work at `create` time? I want to be able to change the dns server used on an existing machine (`--engine-arbitrary-flag dns=8.8.8.8 --engine-arbitrary-flag dns-search=.`).

A machine restart is okay. I do this now by editing EXTRA_ARGS manually and restarting.

nathanleclaire commented:
> Does this feature only work at create time?

Yes -- for now at least.  I think eventually we would like to support mutating machine options "in flight", but that's probably a long way off.

> A machine restart is okay. I do this now by editing EXTRA_ARGS manually and restarting.

Yep, that's the gist of how I would do that with an existing machine.

SvenDowideit commented:
much +1 from here

nathanleclaire commented on libmachine/provision/utils.go:
```diff
@@ -166,16 +173,12 @@ func ConfigureAuth(p Provisioner, authOptions auth.AuthOptions) error {
-	return nil
-}
+	// TODO: Do not hardcode daemon port, ask the driver
+	if err := utils.WaitForDocker(ip, dockerPort); err != nil {
```
This fixes a race which was causing issues in the integration tests - basically, there is a small window after a create finishes where the docker daemon is still starting up.  It's really hard to notice if doing things by hand but if you want to run a `docker $(docker-machine config host) ps` type command _immediately_ after `docker-machine create` you're gonna have a bad time without this.

ehazlett replied:
+1

nathanleclaire commented:
Update: I've added a few integration tests to test this functionality.  I've run into some occasional issues running them on cloud providers but after fixing a race condition related to daemon start they mostly seem to be related to other code.

ehazlett commented:
nice -- ready to get this in -- i will test it today :D

ehazlett commented:
LGTM

zrml commented:
Ideally it would be nice to have this fix, work out what is needed when I provision the standard boot2docker image and I specify:
--engine-storage-driver devicemapper
and if the lvm2 packaged needs to be installed (like for Ubuntu) or other things just do it for me.

Thanks
